### PR TITLE
feat: introduce helm hook `pre-upgrade`

### PR DESCRIPTION
**Pull request by mantissahz**
Add helm hook `pre-upgrade` file `preupgrade-job.yaml` to protect the system from unsupported upgrade path.

longhorn/longhorn#5131

### CODE REVIEW

**Comment by innobead** [comment]
@mantissahz is this ready for review?

**Comment by mantissahz** [comment]
@innobead LEP https://github.com/longhorn/longhorn/pull/5583 is not approved yet

**Comment by innobead** [comment]
@mantissahz https://github.com/longhorn/longhorn/pull/5583#pullrequestreview-1374813656

**Comment by innobead** [comment]
ready?

**Comment by mantissahz** [comment]
This PR should wait for https://github.com/longhorn/longhorn-manager/pull/1763 because the new command `pre-upgrade` of `longhorn-manager` will be introduced.

File: chart/templates/preupgrade-job.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    metadata:
+      name: longhorn-pre-upgrade
+      labels: {{- include "longhorn.labels" . | nindent 8 }}
+    spec:
+      containers:
+      - name: longhorn-post-upgrade
+        image: {{ template "registry_url" . }}{{ .Values.image.longhorn.manager.repository }}:{{ .Values.image.longhorn.manager.tag }}
+        imagePullPolicy: {{ .Values.image.pullPolicy }}
+        securityContext:
+          privileged: true
```
Comment (innobead) [inline chart/templates/preupgrade-job.yaml:23]: Why need privileged?

**Review by innobead — CHANGES_REQUESTED** [review]

File: chart/templates/preupgrade-job.yaml (hunk 1 above)
Comment (mantissahz) [reply chart/templates/preupgrade-job.yaml:23]: It's learned from `post-upgrade` and `uninstall` but it seems we don't need privileged if we change nothing.

Comment (innobead) [reply chart/templates/preupgrade-job.yaml:23]: yeah, make sure the reason for that is.

@shuo-wu @PhanLe1010 @derekbit do you know why we need `privileged` for `post-upgrade` and `uninstall` job?

Comment (PhanLe1010) [reply chart/templates/preupgrade-job.yaml:23]: Good question. It was enabled by this commit https://github.com/longhorn/longhorn/commit/eb767033390cc31c2a01ec3f93dfd2eb3ff05b1a

But I just walk through the post-upgrade and uninstall job. They are just doing operation against kubeAPI. Not sure why we need them to be privileged

Comment (innobead) [reply chart/templates/preupgrade-job.yaml:23]: Let's work on https://github.com/longhorn/longhorn/issues/5862 @mantissahz

Comment (mantissahz) [reply chart/templates/preupgrade-job.yaml]: Removed this setting first. Thanks

**Review by innobead — APPROVED** [review]
LGTM. 

Will merge this until the manager PR is merged.

File: chart/templates/preupgrade-job.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+spec:
+  activeDeadlineSeconds: 900
+  backoffLimit: 1
+  template:
+    metadata:
+      name: longhorn-pre-upgrade
+      labels: {{- include "longhorn.labels" . | nindent 8 }}
+    spec:
+      containers:
+      - name: longhorn-post-upgrade
```
Comment (shuo-wu) [inline chart/templates/preupgrade-job.yaml:19]: NIT: Why is the name `longhorn-post-upgrade`?

**Comment by mantissahz** [comment]
@innobead 
fixed, https://github.com/longhorn/longhorn/pull/5996

**Pull request merged**